docs(guides): add runnable E2B code evaluation guide

Ayush7614 opened the pull request:
## Summary

- Add a concise guide for evaluating LLM-generated Python through an E2B sandbox.
- Merge current `origin/main` into the fork branch (`f8f50a7a3`).
- Correct the example to install the current published E2B SDK, `e2b-code-interpreter==2.7.0`.
- Strengthen the runnable validator so each generated function must pass several hidden cases instead of one answer (`e6291c633`).

## Audit Repairs

- Prior audit text stated that E2B `2.7.0` was unavailable and pinned `2.6.2`. Authoritative PyPI metadata now reports `2.7.0`, uploaded on 2026-05-22, as the current published release. The guide and example README now install that verified release.
- Installed `e2b-code-interpreter==2.7.0` and inspected its API: `Sandbox.create` accepts `allow_internet_access=False`, and `Sandbox.run_code` accepts `language="python"` and `timeout=5`, so the documented isolation and timeout calls remain valid.
- The example previously accepted each generated function after one hidden input/output check. A constant implementation such as `def factorial(n): return 120` therefore passed its advertised factorial evaluation. `test_cases` now executes three hidden cases in the same sandbox run for each generated function, while preserving the singular-variable fallback for existing configurations.
- Updated the guide and README to describe the multi-case contract and show how to configure it.
- Retained the earlier repairs for sandbox provisioning failures, portable metrics filenames, supported SDK calls, and concise example/report documentation.

## Validation

- Queried PyPI package metadata and installed `e2b-code-interpreter==2.7.0` in a disposable virtual environment; inspected `Sandbox.create` and `Sandbox.run_code` signatures.
- `ruff check --select F401,F841,I --fix ...` and `ruff format --check ...` using CI-pinned Ruff `0.15.1`.
- `python -m py_compile` for `validate_and_run_code_e2b.py`, `metrics.py`, `report.py`, and the deterministic assertion harness.
- Deterministic assertion harness verifies correct multi-case output, rejection of the former constant-answer bypass, network-disabled sandbox creation, five-second execution timeout forwarding, unsafe-pattern rejection, and provisioning-error reporting.
- `npm run f`, `npm run l`, `npx prettier --check examples/integration-e2b/README.md examples/integration-e2b/promptfooconfig.yaml site/docs/guides/evaluate-e2b.md`, and `git diff --check`.
- `npx vitest run test/examples/readme.test.ts --sequence.shuffle=false` (1 file, 1669 tests passed).
- `SKIP_OG_GENERATION=true npm run build` from `site/` (passed).
- `npm test` (720 files passed, 1 skipped; 18,311 tests passed, 10 skipped).
- Real promptfoo eval of the checked-in YAML with deterministic E2B/provider adapters:
  - `eval-YVD-2026-05-26T13:23:57`: valid implementations passed all 3 tasks and all 9 hidden checks.
  - `eval-po4-2026-05-26T13:24:04`: the deliberately hard-coded factorial was rejected (`Expected 1 / 120 / 5040, got 120 / 120 / 120`), while the two genuine implementations passed.
- Fresh GitHub Actions on final head `e6291c633` completed with 29 successful checks, 2 intentional skips, and no failures.

Live hosted E2B and OpenAI requests were not made because they require contributor credentials; the real promptfoo eval uses deterministic adapters to verify the documented assertion, policy, and scoring contract.

Ayush7614 committed:
docs: add e2b-code-eval-sandbox full fledged guide

addelong reviewed (changes requested):
Thank you very much for the e2b guide Ayush! I apologize for the delay in reviewing this PR, we don't like leaving our contributors waiting that long and that's on me. I've left some feedback for you on the doc. In general, I would love if this guide could match the tone and structure of other existing guides a bit more - currently it's a long setup section and then several pages of Python code. Reads more like internal engineering notes than a user-facing guide. Would prefer a more concise setup section, links to Python files elsewhere in the repo, and more screenshots and narrative about _why_ these things matter to the user.

Thank you again for your contribution! I look forward to reviewing it in a more timely manner once the edits are made.

Ayush7614 commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,983 @@
+- In this step, we’ll define all the files needed for evaluating LLM-generated code securely with e2b sandbox.
+- This includes validation logic, metrics logging, report generation, YAML config, and prompt templates.
+
+### Create `validate_and_run_code_e2b.py`
```
Done

Ayush7614 commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,983 @@
+- Assertion results: Detailed breakdown of each check run via validate_and_run_code_e2b.py, including sandbox errors, safety violations, or mismatched outputs.
+- This makes it easy to see not just what passed or failed, but also why—with access to sandbox logs and execution details.
+
+### Step 9: Set Up Red Team Target (Custom E2B Provider)
```
Done

Ayush7614 committed:
updated guide & complete feedback

Ayush7614 commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,983 @@
+      problem: "Write a Python function to find the largest element in a list"
+      function_name: "find_largest"
+      test_input: "[1, 5, 3, 9, 2]"
+      expected_output: "9"
```
Done

Ayush7614 replied:
Done

Ayush7614 commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,983 @@
+def model_call_patch(fail_log: str, relevant_files: dict) -> str:
+    # Placeholder function: should call your LLM provider / promptfoo to get a patch (unified diff string)
+    # Return: unified diff (git apply-ready)
+    raise NotImplementedError("Implement model_call_patch()")
```
Done

Ayush7614 committed:
fix images and guide

Ayush7614 committed:
docs: add evaluate-e2b guide with demo GIF and results screenshot

Ayush7614 commented:
cc: @addelong 

## PR Change Summary

- Refined the conclusion to clearly state the value proposition and added a numbered flow (Promptfoo → LLM → e2b sandbox → results) for better clarity.

- Moved the lengthy validate_and_run_code_e2b.py script to the /examples directory and linked it in the guide to improve readability.

- Summarized metrics.py and report.py sections in the guide and linked out to the full files to keep the documentation concise and readable.

- Replaced the NotImplementedError placeholder with a minimal working example in model_call_patch to demonstrate how to call the LLM and generate a patch.

- Added an additional test case to demonstrate detection of unsafe code and handling of failing scenarios, improving coverage and realism.

- Adjusted the “Step 9” heading level to match the formatting of other steps for consistency and clarity.

- Added relevant screenshots (table view and web viewer dashboard) to visually demonstrate test results and improve guide clarity.

- Expanded the final section to elaborate on advanced enhancements like unsafe pattern detection, patch workflows, key management, and automated red teaming.

In addition, the e2b-code-eval example already existed under the examples directory in Promptfoo, and I extended and completed it as part of this PR. I linked all the relevant code files (validate_and_run_code_e2b.py, metrics.py, swe_runner.py, etc.) directly from the examples to the updated guide, ensuring better navigation and readability. During the review process, several improvements were also suggested in the example code itself — such as refining test cases, adding a failing/unsafe scenario, and replacing the NotImplementedError placeholder in model_call_patch with a minimal working example. Since these changes involve modifying the example source files, I created a new PR to address those code-level improvements separately from the documentation updates.

PR Link: https://github.com/promptfoo/promptfoo/pull/5753
#5753

addelong reviewed (approved):
Thank you for the changes @Ayush7614! I've approved both PRs, in the future feel free to just keep everything in one PR.

Appreciate the contribution!

Ayush7614 commented:
CC @mldangelo @typpo

nidhishgajjar commented:
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))

## Summary

This PR adds comprehensive documentation for evaluating LLM code generation using Promptfoo with E2B sandbox integration. The guide covers the complete workflow from initial setup through red teaming and security analysis, including how to safely execute model-generated code in an isolated sandbox environment.

## Architecture

This is a documentation-only PR with three new files:

1. **Main documentation guide** (`site/docs/guides/evaluate-e2b.md`):
   - Complete 13-step tutorial for E2B + Promptfoo integration
   - Covers environment setup, project initialization, code execution in sandbox
   - Includes red teaming setup for security testing
   - Provides detailed explanations and code examples

2. **Supporting assets**:
   - `site/static/img/docs/e2b/e2b-code.gif`: Animated demo of secure code generation
   - `site/static/img/docs/e2b/e2b-tests.png`: Screenshot of Promptfoo test results view


3. **Content structure**:
   - Step-by-step instructions from environment setup through final results analysis
   - Includes practical code examples for validation scripts and configuration
   - Shows integration with both standard evaluation and red team workflows
   - Provides security-focused guidance for safe code evaluation


## Issues

**Documentation quality concerns** (`site/docs/guides/evaluate-e2b.md`) — [comment] — The guide is comprehensive but has several issues that could improve clarity:

- Inconsistent formatting: Some sections use mixed formatting (e.g., "What's happening here?" vs "Why do we need these?")
- Typos and grammar issues: Several instances throughout (e.g., "Collecting" instead of "Collecting", "networking" instead of "network access")
- Overly verbose: Some explanations are wordy and could be more concise


**Missing technical details** (`site/docs/guides/evaluate-e2b.md`) — [suggestion] — The guide mentions important scripts but doesn't provide:
- Actual implementation details for `validate_and_run_code_e2b.py` safety checks
- Specific regex patterns used for dangerous code detection
- Performance characteristics and limitations of E2B sandbox
- Cost considerations for repeated evaluations

**Incomplete reference to example files** (`site/docs/guides/evaluate-e2b.md`) — [comment] — The guide extensively references example files (validate_and_run_code_e2b.py, metrics.py, etc.) but these files don't appear to be included in the PR. Readers will get 404 errors when clicking the GitHub links unless:
- These files exist in the repository
- The PR should include them or link to where they will be added
- Or the guide should provide inline code snippets instead

**Missing troubleshooting section** — [suggestion] — No troubleshooting guide for common issues:
- What if E2B sandbox times out or fails to start?
- How to handle API rate limits from OpenAI?
- Common errors in YAML configuration
- Debugging when tests fail unexpectedly

**Limited scope considerations** (`site/docs/guides/evaluate-e2b.md`) — [comment] — The guide only covers Python evaluation, but:
- No mention of JavaScript code execution in E2B
- No guidance on multi-language projects
- No discussion of alternative sandbox options (Docker, etc.)

## Cross-file Impact

- `site/docs/guides/evaluate-e2b.md`: New comprehensive tutorial
- `site/static/img/docs/e2b/`: Two new documentation assets (animated demo and screenshot)
- Promptfoo documentation: Adds significant new content for security-focused code evaluation
- Users: Will have clear guidance on setting up secure LLM code evaluation workflows

## Assessment

⚠️ Request changes

This PR adds valuable documentation content, but there are several issues that need addressing:

1. **Missing example files**: The guide extensively references scripts (validate_and_run_code_e2b.py, metrics.py, report.py, swe_runner.py, etc.) but these files are not included in the PR. Readers following the guide will not be able to complete it without these files.

2. **Quality issues**: Multiple typos, formatting inconsistencies, and overly verbose explanations that reduce readability.

3. **Incomplete technical details**: Missing critical implementation details (regex patterns for safety checks, specific error handling) that are necessary for users to actually implement the workflow.

4. **No troubleshooting**: No guidance on common issues users will encounter.

The content is well-structured and covers an important topic (secure LLM code evaluation), but it needs to either:
- Include all referenced example files in this PR, or
- Make clear these files are part of a separate upcoming PR, or
- Provide inline code snippets so the guide is self-contained

Without these fixes, the documentation will be frustrating for users who try to follow it and hit missing files.

mldangelo-oai committed:
Merge branch 'main' into ayush5

mldangelo-oai committed:
Merge branch 'main' into ayush5

mldangelo-oai committed:
Merge branch 'main' into ayush5

mldangelo-oai committed:
Merge branch 'main' into ayush5

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into ayush5

mldangelo-oai committed:
fix(docs): make E2B guide enforce documented isolation

mldangelo-oai commented on examples/grok-4-vs-gpt-5/promptfooconfig.yaml:
```diff
@@ -0,0 +1,126 @@
+      temperature: 0.7
+      max_tokens: 4000
+
+prompts:
```
Resolved as superseded on delivered head c64393d41: examples/grok-4-vs-gpt-5/promptfooconfig.yaml is not in the current PR diff.

mldangelo-oai commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,983 @@
+- Validate outputs against test cases.
+- Collect metrics (runtime, success/failure, logs).
+
+By the end of this guide, you’ll have a working setup where Promptfoo calls e2b for sandboxed execution, runs automated tests, and produces reproducible evaluation reports.
```
Addressed in c64393d41: the rewritten guide opens with the sandboxed evaluation value and includes a numbered Promptfoo-to-provider-to-E2B-to-result execution flow.

mldangelo-oai commented on examples/grok-4-vs-gpt-5/promptfooconfig.yaml:
```diff
@@ -0,0 +1,126 @@
+  - name: 'Semantic Quality'
+    metric: semantic_similarity
+    reference: 'Evaluate correctness, reasoning, coding quality, refusal behavior, bias handling, and overall coherence'
+
```
Resolved as superseded on delivered head c64393d41: examples/grok-4-vs-gpt-5/promptfooconfig.yaml is not in the current PR diff.

mldangelo-oai commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,962 @@
+
+    # Last resort: just run code with defaults
+    return sbx.run_code(code)
+
```
Addressed in c64393d41 after verifying e2b-code-interpreter 2.7.0 signatures: the example now creates Sandbox with allow_internet_access=False and calls run_code with language="python" and timeout=5 only; unsupported limits and allow_network fallbacks are removed.

mldangelo-oai commented on examples/grok-4-vs-gpt-5/README.md:
````diff
@@ -0,0 +1,38 @@
+
+```bash
+promptfoo view
+```
````
Resolved as superseded on delivered head c64393d41: examples/grok-4-vs-gpt-5/README.md is not in the current PR diff.

mldangelo-oai commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,983 @@
+
+Type y, and your browser will open the Promptfoo dashboard.
+
+## What you see in the Promptfoo Web Viewer:
```
Disposition in c64393d41: the uploaded screenshot was removed because it showed a superseded gpt-4.1 six-case run and contradicted the runnable config. The guide now directs readers to inspect their current exported results.

mldangelo-oai commented on site/docs/guides/evaluate-e2b.md:
```diff
@@ -0,0 +1,983 @@
+- Extend unsafe pattern detection (cover more dangerous libraries).
+- Add patch generation workflows with swe_runner.py.
+- Use .env.example for easier key management.
+- Automate daily red team runs with GitHub Actions.
```
Addressed in c64393d41: the rewritten guide explains sandbox policy, customization, and the correct boundary for red teaming an application target.

mldangelo-oai commented:
@codex review

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into mdangelo/codex/audit-pr-5555

mldangelo-oai committed:
fix(docs): make E2B evaluation example runnable

mldangelo-oai committed:
style(docs): format E2B assertion example

mldangelo-oai committed:
Merge remote-tracking branch 'origin/main' into mdangelo/codex/audit-pr-5555

mldangelo-oai committed:
docs(site): strengthen E2B code evaluation example
